drivers/sdmmc/sdmmc_sdhc.c

```diff
@@ -0,0 +1,900 @@
+    uint32_t now = start;
+
+    _ZTIMER_ACQUIRE();
+    while ((sdhc->PSR.reg & (SDHC_PSR_CMDINHC | SDHC_PSR_CMDINHD)) &&
+           ((now - start) < SDHC_BUSY_TIMEOUT)) {
+        now = _ZTIMER_NOW();
+    }
+    _ZTIMER_RELEASE();
+    return ((now - start) >= SDHC_BUSY_TIMEOUT) ? -ETIMEDOUT : 0;
+}
+
+/**
+ * @brief   Wait for a given event while checking for errors
+ *
+ * @param   state       SDHC device context
+ * @param   event       Event to wait for [SDHC_NISTR_*]
+ * @param   error_mask  Mask of errors to be checked [SDHC_EISTR_*]
+ * @param   reset       Reset type in case of errors
+ *                      [SDHC_SRR_SWRSTALL | SDHC_SRR_SWRSTCMD | SDHC_SRR_SWRSTDAT]
+ *
+ * @return  true if the given event has occurred, or false if an error has occurred.
+ */
+static bool _wait_for_event(sdhc_dev_t *sdhc_dev,
+                            uint16_t event, uint16_t error_mask,
+                            uint8_t reset)
+{
+    sdhc_t *sdhc = sdhc_dev->conf->sdhc;
+
+    sdhc_dev->error = 0;
+
+    /* TODO: timeout */
+    sdhc->NISIER.reg |= event;
+    sdhc->EISIER.reg |= error_mask;
+
+    /* block IDLE so that the CPU clock does not stop */
+
+#if defined(MCU_SAMD5X) || defined(MCU_SAME5X)
+    pm_block(SAM0_PM_IDLE);
+#endif
+    mutex_lock(&sdhc_dev->irq_wait);
+#if defined(MCU_SAMD5X) || defined(MCU_SAME5X)
+    pm_unblock(SAM0_PM_IDLE);
+#endif
+
+    sdhc->NISIER.reg &= ~event;
+    sdhc->EISIER.reg &= ~error_mask;
+
+    if (sdhc_dev->error & error_mask) {
+        if (IS_USED(ENABLE_DEBUG)) {
+            DEBUG("[sdmmc] SDHC error: EISTR=%04x, ", sdhc_dev->error);
+            switch (reset) {
+            case SDHC_SRR_SWRSTCMD:
+                DEBUG("reset CMD\n");
+                break;
+            case SDHC_SRR_SWRSTDAT:
+                DEBUG("reset DAT\n");
+                break;
+            case SDHC_SRR_SWRSTALL:
+                DEBUG("reset ALL\n");
+                break;
+            default:
+                assert(false);
+                break;
+            }
+        }
+        _reset_sdhc(sdhc_dev, reset);
+        return false;
+    }
+
+    return true;
+}
+
+static int _enable_sd_clk(sdhc_dev_t *sdhc_dev)
+{
+    sdhc_t *sdhc = sdhc_dev->conf->sdhc;
+
+    if (!(sdhc->CCR.reg & SDHC_CCR_SDCLKEN)) {
+        DEBUG("[sdmmc] enable SDCLK\n");
+        /* TODO timeout handling */
+        sdhc->CCR.reg |= SDHC_CCR_SDCLKEN;   /* enable clock to card        */
+
+        /* a very small delay is required after clock changing */
+        volatile unsigned count = (CLOCK_CORECLOCK / US_PER_SEC) * 10;
+        while (--count) {}
+    }
+
+    return 0;
+}
+
+static int _disable_sd_clk(sdhc_dev_t *sdhc_dev)
+{
+    sdhc_t *sdhc = sdhc_dev->conf->sdhc;
+
+    /* TODO timeout handling */
+    if (sdhc->CCR.bit.SDCLKEN) {
+        DEBUG("[sdmmc] disable SDCLK\n");
+        /* wait for command/data to go inactive */
+        while (sdhc->PSR.reg & (SDHC_PSR_CMDINHC | SDHC_PSR_CMDINHD)) {}
```

Comment:
Unfortunately I really *do* see it getting stuck here when I run `tests/sys/vfs_default` with

    USEMODULE += lwext4_vfs
    CFLAGS += -DCONFIG_MTD_SDMMC_ERASE=1 # not sure if this is really needed

The test will start by formatting the SD card (which seems to be done rather inefficiently, there are many duplicate writes it seems) but this onslaught of SD commands does a good job of triggering the bug.

I tried adding a naive timeout:

```patch
--- a/drivers/sdmmc/sdmmc_sdhc.c
+++ b/drivers/sdmmc/sdmmc_sdhc.c
@@ -805,7 +805,14 @@ static int _disable_sd_clk(sdhc_dev_t *sdhc_dev)
     if (sdhc->CCR.bit.SDCLKEN) {
         DEBUG("[sdmmc] disable SDCLK\n");
         /* wait for command/data to go inactive */
-        while (sdhc->PSR.reg & (SDHC_PSR_CMDINHC | SDHC_PSR_CMDINHD)) {}
+        uint32_t timeout = 0xfffff;
+        while (sdhc->PSR.reg & (SDHC_PSR_CMDINHC | SDHC_PSR_CMDINHD)) {
+            if (--timeout == 0) {
+                break;
+            }
+        }
+        printf("cmd took %lu loops\n", 0xfffff - timeout);
+
         /* disable the clock to card */
         sdhc->CCR.reg &= ~SDHC_CCR_SDCLKEN;
     }
```

but the controller does not recover from this condition:

```
2023-11-07 19:02:20,637 - INFO # mtd_sdmmc_write_page: page:442384 offset:0 size:1024
2023-11-07 19:02:20,643 - INFO # cmd took 0 loops
2023-11-07 19:02:20,645 - INFO # write bgroup 36 / 948
2023-11-07 19:02:20,649 - INFO # lwext4: erase 2 sectors starting with lu
2023-11-07 19:02:20,652 - INFO # cmd took 0 loops
2023-11-07 19:02:20,655 - INFO # lwext4: write 1024 bytes to page 458766
2023-11-07 19:02:20,660 - INFO # mtd_sdmmc_write_page: page:458766 offset:0 size:1024
2023-11-07 19:02:20,666 - INFO # cmd took 0 loops
2023-11-07 19:02:20,667 - INFO # write bgroup 37 / 948
2023-11-07 19:02:20,671 - INFO # lwext4: erase 2 sectors starting with lu
2023-11-07 19:02:22,980 - INFO # cmd took 1048575 loops
2023-11-07 19:02:22,984 - INFO # lwext4: write 1024 bytes to page 475150
2023-11-07 19:02:22,988 - INFO # mtd_sdmmc_write_page: page:475150 offset:0 size:1024
2023-11-07 19:02:23,490 - INFO # [sdmmc] Card not present
2023-11-07 19:02:23,493 - INFO # mtd_sdmmc_write_page: error -19
```